plugin-server/src/main/ingestion-queues/session-recording/blob-ingester/session-manager.ts

```diff
@@ -380,30 +410,39 @@ export class SessionManager {
         // If it is a chunked message we add to the collected chunks
 
         if (!this.chunks.has(message.chunk_id)) {
-            this.chunks.set(message.chunk_id, [])
+            this.chunks.set(message.chunk_id, new PendingChunks(message))
+        } else {
+            this.chunks.get(message.chunk_id)?.add(message)
         }
-        const chunks: IncomingRecordingMessage[] = this.chunks.get(message.chunk_id) || []
-        chunks.push(message)
+        const pendingChunks = this.chunks.get(message.chunk_id)
 
-        if (chunks.length === message.chunk_count) {
+        if (pendingChunks && pendingChunks.isComplete) {
             // If we have all the chunks, we can add the message to the buffer
             // We want to add all the chunk offsets as well so that they are tracked correctly
-            chunks.forEach((x) => {
-                this.buffer.offsets.push(x.metadata.offset)
-            })
-
-            await this.addToBuffer({
-                ...message,
-                data: chunks
-                    .sort((a, b) => a.chunk_index - b.chunk_index)
-                    .map((c) => c.data)
-                    .join(''),
-            })
-
+            gaugePendingChunksCompleted.inc()
+            await this.processChunksToBuffer(pendingChunks.completedChunks)
             this.chunks.delete(message.chunk_id)
         }
     }
 
+    private async processChunksToBuffer(chunks: IncomingRecordingMessage[]) {
+        await this.addToBuffer({
+            ...chunks[chunks.length - 1],
+            data: chunks
+                .sort((a, b) => a.chunk_index - b.chunk_index)
+                .map((c) => c.data)
+                .join(''),
+        })
+        // push all but the last offset into the buffer
+        // the final offset was copied into the data passed to `addToBuffer`
+        for (let i = 0; i < chunks.length - 1; i++) {
+            const x = chunks[i]
+            this.buffer.offsets.push(x.metadata.offset)
+        }
```

Comment:
We're not relying on order any more so changed it back but in case it matters...

* the flush on size if after this awaited `addToChunks` so they don't clash
* this isn't inside a try/catch so if we can't e.g. decompress we throw all the way out

So, we aren't relying on the order of these two operations (I don't think) but if we care we should wrap it in a test or code so we can't flip the order and cause a break in future